### PR TITLE
gossiping_property_file_snitch_test: modernize + fix potential race

### PR DESCRIPTION
**Pull request by gusev-p**
This is mostly a refactoring PR to make the test more readable, as a byproduct of
scylladb/scylladb#17369 investigation.

We add the check for specific type of exceptions that can be thrown (`bad_property_file_error`).

We also fix the potential race - the test may write to res from multiple cores with no locks.

### CODE REVIEW

File: test/boost/gossiping_property_file_snitch_test.cc [hunk 1]
```diff
@@ -15,78 +15,50 @@
...
-                        return make_ready_future<>();
-                    });
-                });
-            } catch (std::exception& e) {
-                BOOST_CHECK(!exp_result);
-                return make_ready_future<>();
-            }
-        }).finally([ snitch_i = std::move(snitch_i) ] () mutable {
-            return snitch_i->stop().finally([snitch_i = std::move(snitch_i)] {});
+    {
```
Comment (xemul) [inline test/boost/gossiping_property_file_snitch_test.cc:51]: If exp_result is false this block is not supposed to be called, but in new implementation is seems to be called.

Comment (gusev-p) [reply test/boost/gossiping_property_file_snitch_test.cc]: yes, fixed

Comment (xemul) [reply test/boost/gossiping_property_file_snitch_test.cc]: Silly question. The test got ... broken, but still passed. Does the test make any sense at all?

Comment (gusev-p) [reply test/boost/gossiping_property_file_snitch_test.cc]: why broken? we just checked that `snitch_base.{_my_dc/_my_rack}` are empty on all shards if we failed to parse the config.

Comment (xemul) [reply test/boost/gossiping_property_file_snitch_test.cc]: :smile:  OK, makes sense. Thanks

**Comment by gusev-p** [comment]
@xemul

**Pull request closed without merging**